### PR TITLE
[RLlib] Fix attention net test for connectors

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Signed-off-by: Artur Niederfahrenhorst <artur@anyscale.com>

## Why are these changes needed?

Attention net wrappers don't provide get_initial_state for agent_collector to build state internally.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by gjoliver — APPROVED** [review]
wow, this is a nice pr.

**Review by kouroshHakha — COMMENTED** [review]
Left some comments regarding the proposed fixes.

File: rllib/models/tf/attention_net.py [hunk 1]
```diff
@@ -918,3 +914,8 @@ def call(
             memory_outs,
             {SampleBatch.VF_PREDS: tf.reshape(value_outs, [-1])},
         )
+
+    @override(ModelV2)
+    def get_initial_state(self) -> Union[List[np.ndarray], List[TensorType]]:
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/attention_net.py:926]: Thanks! Left a comment

File: rllib/examples/attention_net.py [hunk 2]
```diff
@@ -193,26 +191,14 @@ def get_cli_args():
...
-            attention_dim = config["model"]["attention_dim"]
-            memory = config["model"]["attention_memory_inference"]
-            init_state = state = [
-                np.zeros([memory, attention_dim], np.float32)
-                for _ in range(num_transformers)
-            ]
             # run one iteration until done
             print(f"RepeatAfterMeEnv with {config['env_config']}")
             while not done:
-                action, state_out, _ = algo.compute_single_action(obs, state)
```
Comment (kouroshHakha) [inline rllib/examples/attention_net.py:207]: I don't think this is right. compute_single_action(obs) won't keep track of the states for you. You'd still have to manually handle state_out and passing it in as the new state. For inti_state you can use `policy.get_initial_state()` and go with that one. To see the error you can run `python rllib/examples/attention_net.py --as-test --stop-reward=70 --framework tf --no-tune` both on master and on this branch. It shows reward of greater than 70 on master, but not on this branch.

Comment (kouroshHakha) [reply rllib/examples/attention_net.py:207]: Also this code-path is not invoked in CI as of now.

File: rllib/models/tf/attention_net.py [hunk 3]
```diff
@@ -549,10 +549,6 @@ def forward(
         model_out = self._logits_branch(self._features)
         return model_out, memory_outs
 
-    @override(ModelV2)
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/models/tf/attention_net.py:552]: I now get why you are deleting it, because the super() class has the same implementation. But I still think this should be overriden to explicitly return np.zeros(). The model creation stack does some weird magic with the Wrappers to "make it work" which is not transparent at all. Explicitly overriding it here would solve that issue imo.

File: rllib/models/tf/attention_net.py [hunk 4]
```diff
@@ -549,15 +549,18 @@ def forward(
...
-    def get_initial_state(self) -> Union[List[np.ndarray], List[TensorType]]:
-        return []
-
     @override(ModelV2)
     def value_function(self) -> TensorType:
         assert self._features is not None, "Must call forward() first!"
         return tf.reshape(self._value_branch(self._features), [-1])
 
+    @override(ModelV2)
```
Comment (kouroshHakha) [inline rllib/models/tf/attention_net.py:557]: ok you are doing it here :)

File: rllib/models/tf/attention_net.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/models/tf/attention_net.py:552]: You are doing what I suggested a few lines below. Good! :)

File: rllib/evaluation/collectors/agent_collector.py [hunk 5]
```diff
@@ -305,12 +314,24 @@ def add_action_reward_next_obs(self, input_values: Dict[str, TensorType]) -> Non
             ):
                 if should_flatten_action_key:
                     v = flatten_to_single_ndarray(v)
+                # Briefly remove dummy value to add to buffer
+                if k in self.gained_from_inference:
+                    dummy = self.buffers[k][0].pop(-1)
```
Comment (kouroshHakha) [inline rllib/evaluation/collectors/agent_collector.py:319]: some note for code optimization nerds like me :) pop(-1) takes O(1) in python lists but pop(0) takes O(n). So we are good here

File: rllib/evaluation/collectors/agent_collector.py [hunk 6]
```diff
@@ -116,6 +116,15 @@ def __init__(
...
         self.agent_steps = 0
+        # Keep track of view requirements that have a view on columns that we gain from
+        # inference and also need for inference. These have dummy values prepended in
+        # buffers to account for the missing value when building for inference
+        # Example: We have one 'state_in' view requirement that has a view on our
+        # state_outs at t=[-10, ..., -1]. At any given build_for_inference()-call,
+        # the buffer must contain eleven values from t=[-10, ..., 0] for us to index
+        # properly. Since state_out at t=0 is missing, we substitute it with a buffer
+        # value that should never make it into batches built for training.
+        self.gained_from_inference = set()
```
Comment (kouroshHakha) [inline rllib/evaluation/collectors/agent_collector.py:127]: Suggestion: I personally will forget what `gained_from_inference` actually means in the future. I think a better name would be something like `data_cols_with_dummy_values`. We also need to use this opportunity, while we have fresh memory about the details, to update the main docstring of the agent_collector class to explain the content of the buffer and how the dummy value will be treated during both inference and training. I'll write a draft when the other matters are resolved.

**Review by kouroshHakha — COMMENTED** [review]
The PR looks good to me, except the changes made to `examples/attention.py`. We also need to run release tests on this PR before we merge. Specially the multi-gpu tests with attention and the learning ones, I wanna make sure we are not breaking anything related to common use-cases. Also `test_agent` ci test is failing.

File: rllib/models/tf/attention_net.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/models/tf/attention_net.py:552]: Sure thing!

File: rllib/evaluation/collectors/agent_collector.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/collectors/agent_collector.py:319]: Oh there is so much to optimize here 😄

File: rllib/examples/attention_net.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/examples/attention_net.py:207]: I forgot that this is different code than with the input_from_raw_dict methods now. We should change this very soon.

Comment (kouroshHakha) [reply rllib/examples/attention_net.py:207]: yep, but in a different PR.

File: rllib/connectors/tests/test_agent.py [hunk 7]
```diff
@@ -430,10 +430,9 @@ def test_vr_connector_with_multiple_buffers(self):
             # next state and action at time t-1 are the following
             timestep_data = {
                 SampleBatch.NEXT_OBS: obs_arrs[t],
-                SampleBatch.ACTIONS: (
-                    np.zeros_like(act_arrs[0]) if t == 0 else act_arrs[t - 1]
-                ),
             }
+            if t > 0:
```
Comment (ArturNiederfahrenhorst) [inline rllib/connectors/tests/test_agent.py:434]: This is just to make it clear that this will not be used on the "0th" timestep anyway. The first timeste_data is only the initial obs that initializes the agent collector inside ViewRequirementsConnector.

File: rllib/evaluation/collectors/agent_collector.py [hunk 8]
```diff
@@ -448,14 +470,20 @@ def build_for_training(
...
+                # value at the last position of the buffer that makes it one longer).
                 count = int(
                     math.ceil(
-                        (len(d) - self.shift_before) / view_req.batch_repeat_value
+                        (
+                            len(d)
+                            - int(data_col in self.gained_from_inference)
+                            - self.shift_before
+                        )
+                        / view_req.batch_repeat_value
```
Comment (ArturNiederfahrenhorst) [inline rllib/evaluation/collectors/agent_collector.py:483]: There is going to be a follow up that makes it clearer what is happening here in the docs.
Most importantly: During inference some values for the current timestep are missing when building the batch. We account for this with a dummy value that is disregarded when building for training (where all values are present).

File: rllib/examples/attention_net.py [hunk 9]
```diff
@@ -195,12 +195,7 @@ def get_cli_args():
             total_reward = 0
             # start with all zeros as state
             num_transformers = config["model"]["attention_num_transformer_units"]
-            attention_dim = config["model"]["attention_dim"]
-            memory = config["model"]["attention_memory_inference"]
-            init_state = state = [
-                np.zeros([memory, attention_dim], np.float32)
-                for _ in range(num_transformers)
-            ]
+            state = algo.get_policy().get_initial_state()
```
Comment (ArturNiederfahrenhorst) [inline rllib/examples/attention_net.py:198]: This change is only for consistency of the initial state. Previously, we did not have a proper initial state for the attention wrappers but it did not matter because there was no agent collector that needed to sample from it.

File: rllib/models/torch/attention_net.py [hunk 10]
```diff
@@ -439,7 +439,12 @@ def forward(
 
     @override(ModelV2)
     def get_initial_state(self) -> Union[List[np.ndarray], List[TensorType]]:
-        return []
+        return [
+            torch.zeros(
+                self.gtrxl.view_requirements["state_in_{}".format(i)].space.shape
+            )
+            for i in range(self.gtrxl.num_transformer_units)
+        ]
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/torch/attention_net.py:447]: See https://github.com/ray-project/ray/pull/30626/files#r1035426127

File: rllib/evaluation/collectors/agent_collector.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/evaluation/collectors/agent_collector.py]: As per our discussion: Changed!

File: rllib/evaluation/collectors/agent_collector.py [hunk 11]
```diff
@@ -116,6 +116,15 @@ def __init__(
         # The simple timestep count for this agent. Gets increased by one
         # each time a (non-initial!) observation is added.
         self.agent_steps = 0
+        # Keep track of view requirements that have a view on columns that we gain from
+        # inference and also need for inference. These have dummy values prepended in
```
Comment (kouroshHakha) [inline rllib/evaluation/collectors/agent_collector.py:120]: prepended or appended?

File: rllib/evaluation/collectors/agent_collector.py [hunk 12]
```diff
@@ -655,3 +683,10 @@ def _fill_buffer_with_initial_values(
                 self._build_buffers({data_col: fill_value})
 
         return is_state
+
+    def _prepare_for_gained_during_inference(self, data_col):
+        self.data_cols_with_dummy_values.add(data_col)
+        # For items gained during inference, we prepend a dummy value here so
```
Comment (kouroshHakha) [inline rllib/evaluation/collectors/agent_collector.py:689]: again append :), right?

File: rllib/evaluation/collectors/agent_collector.py [hunk 13]
```diff
@@ -655,3 +683,10 @@ def _fill_buffer_with_initial_values(
                 self._build_buffers({data_col: fill_value})
 
         return is_state
+
+    def _prepare_for_gained_during_inference(self, data_col):
```
Comment (kouroshHakha) [inline rllib/evaluation/collectors/agent_collector.py:687]: nit for consistency: `_prepare_for_data_cols_with_dummy_values`?

**Review by kouroshHakha — APPROVED** [review]
Finally. This is great. Thanks @ArturNiederfahrenhorst. Just a few very minor nits.

File: rllib/connectors/tests/test_agent.py [hunk 14]
```diff
@@ -450,8 +449,13 @@ def test_vr_connector_with_multiple_buffers(self):
...
 
             self.assertTrue("context_next_obs" not in sample_batch)
+            # We should have the 5 (context_len) most recent observations here
             check(sample_batch["context_obs"], np.stack(obs_list)[None])
-            check(sample_batch["context_act"], np.stack(act_list[:-1])[None])
+            # The context for actions is [t-context_len+1:t]. Since we build sample
+            # batch for inference in ViewRequirementAgentConnector, it always
+            # includes everything up until the last action (at t-1), but not the
+            # action current action (at t).
+            check(sample_batch["context_act"], np.stack(act_list[1:])[None])
```
Comment (ArturNiederfahrenhorst) [inline rllib/connectors/tests/test_agent.py:458]: This looked correct before. But the way it was before it never included the action we added before calling build_for_inference. But this is what happens with State Buffer connector.
